java/src/jmri/jmrix/ipocs/IpocsPortController.java

```diff
@@ -0,0 +1,137 @@
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.StandardSocketOptions;
+import java.nio.channels.AsynchronousServerSocketChannel;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import jmri.jmrix.AbstractPortController;
+import jmri.jmrix.ipocs.protocol.Message;
+import jmri.util.zeroconf.ZeroConfService;
+
+/**
+ *
+ * @author Fredrik Elestedt Copyright (C) 2020
+ * @since 4.21.2
+ */
+public class IpocsPortController extends AbstractPortController implements IpocsClientListener {
+  private final static Logger log = LoggerFactory.getLogger(IpocsPortController.class);
+  private static String INADDR_ANY = "0.0.0.0";
+  private AsynchronousServerSocketChannel serverSocket = null;
+  private IpocsSocketAcceptor socketAcceptor;
+  private ZeroConfService zeroConfService = null;
+  private final List<IpocsClientListener> clientListeners = new ArrayList<IpocsClientListener>();
+  private Map<String, IpocsClientHandler> clients = new HashMap<>();
+  private Map<String, Message> lastMessage = new HashMap<>();
+
+  public IpocsPortController(IpocsSystemConnectionMemo memo) {
+    super(memo);
+    super.setManufacturer(IpocsConnectionTypeList.IPOCSMR);
+    final Option o1 = new Option("Listing port", new String[]{"10000"}, false, Option.Type.TEXT);
+    super.options.put(super.option1Name, o1);
+  }
+
+  @Override
+  public IpocsSystemConnectionMemo getSystemConnectionMemo() {
+    return (IpocsSystemConnectionMemo) super.getSystemConnectionMemo();
+  }
+
+  @Override
+  public void configure() {
+    if (getSystemConnectionMemo().getPortController() == null) {
+      getSystemConnectionMemo().setPortController(this);
+    }
+    getSystemConnectionMemo().configureManagers();
+  }
+
+  @Override
+  public void connect() throws IOException {
+    log.info("Setting up service");
+    serverSocket = AsynchronousServerSocketChannel.open();
+    socketAcceptor = new IpocsSocketAcceptor(this, serverSocket);
+    final int port = Integer.parseInt(super.getOptionState(super.option1Name));
+    final InetSocketAddress address = new InetSocketAddress(INADDR_ANY, port);
+    serverSocket.bind(address);
+    serverSocket.setOption(StandardSocketOptions.SO_REUSEADDR, true);
+    serverSocket.accept(null, socketAcceptor);
+    log.info("Starting ZeroConfService _ipocs._tcp.local");
+    zeroConfService = ZeroConfService.create("_ipocs._tcp.local.", "ipocs", 10000, 0, 0, new HashMap<String, String>());
+    zeroConfService.publish();
+  }
+
+  @Override
+  public DataInputStream getInputStream() {
+    // This will never return anything, since this is a server port controller
+    return null;
+  }
+
+  @Override
+  public DataOutputStream getOutputStream() {
+    // This will never return anything, since this is a server port controller
+    return null;
+  }
```

Comment:
Should these methods never be called? If so, it would be better if they throw an UnsupportedOperationException("Not supported"), since that would throw an exception when the method is called and not when the returned stream is used.

Reply:
I opted against that since I did not know if any part of JMRI called these funtions and expected a return or null if the streams should not be used.
If they aren't called by any other part of JMRI - why are they there?

Reply:
Changed